README.md
<!--
Select or linsten bird's voice on R
@author woodiw2wopper <woodie2wopper@gmail.com>
Copyright (C) 2019 woodiw2wopper
-->
# toriR（トリル）｜トリを聴き取る on R (Select or linsten bird's voice on R;  (tori in Japanese is bird) )
- エラーなどうまく動かない場合は、一番下のFAQをお読みください。
## 概要（できること）(What you can do.)
- ICレコーダの録音データから変換されたスペクトログラムを`RStudio`上で表示します。
- 見覚えのある声紋があればクリックし、メニューから種名を選択することで、日時と種名が`csv`ファイルに出力されます。
- 迷う場合はメニューから"再生"を選べば、音が再生できます（`SoX`のインストールが必要）。
- 再生の後にわかれば種名を選び、わからなければ不明種(WAY?: Who Are You?)としてクリック前後の数秒の音を保存できます。
-The spectrogram converted from the recording data of the IC recorder is displayed on `RStudio`.
-Click on a familiar voiceprint and select the species name from the menu. The date and species name will be output to the `csv` file.
-If you are unsure, select "Play" from the menu to play the sound (`SoX` must be installed).
-Select a species name if you know it after playback, and save a few seconds before and after clicking as an unknown species (WAY ?: Who Are You?).

## 対応OS
1. MacOSX
2. Windows10
3. 多分Linux系もOKでしょう

## 環境準備（下記をインストールし、pathに登録してください）
1. `R`
2. `RStudio`
3. `SoX`

## version
| No.  | soft    | version          | link                                                         |
| ---- | ------- | ---------------- | ------------------------------------------------------------ |
| 1    | SoX     | v14.4.2          | [qiita](https://qiita.com/teteyateiya/items/e4dc27e384d947b9946d) |
| 2    | R       | R3.6.1           | [cran.r-project.org](https://cran.r-project.org/)            |
| 3    | RStudio | Version 1.2.5001 | [rstudio](https://www.rstudio.com/)                          |

~編集中~

## 音源データの準備
ファイル名を下記の666形式の録音日時に変更してください。

例えば、2019年05月01日23:00:00から翌日の05月02日02:00:00までの音源データなら
"`開始日にち_開始時刻-終了時刻_xxx.拡張子`”という風に変更してください。”_”と”-“を区別してください。xxxの部分はあってもなくても構いません。

original filename: `DS700143.WMA`
modified filename: `190501_230000-020000_DS700143.WMA`

これを6_6-6形式とあるいは単純に666形式と呼んでいます。
### ディレクトリの構成
```{bash}
$tree -d
.
└── toriR_demo
    ├── 190501
    └── ORG
```    
- ICレコーダの音源はORGというフォルダに入れてください。その親のフォルダに日にち毎のディレクトリが作られて、そこに音声ファイル(.wav)と、スペクトログラム(.png)、設定ファイル(.psgrm)ができます。()は拡張子です。

### Input and output files(入出力ファイル）
`toriR` is based on processing for one hour. For example, if the recording is 1 hour from 23:00:00 on May 01, 2019, the following input files are required:
（toriRは1時間分の処理を基本としており、例えば録音が2019年05月01日23:00:00から1時間の場合、次の入力ファイルが必要です）
```{bash}
Input files:
190501_230000-000000.wav(sound:音声ファイル)
190501_230000-000000.psgrm(parameter:パラメータファイル)
190501_230000-000000_P05.png(spectrogrum for four minutes:4分のスペクトログラム)

~
190501_230000-000000_P15.png
```
Output file is `csv` file
The output file of `toriR` is a `csv` file, and the output information is "date, time, species name and clickpoint frequency" obtained by clicking the spectrogram and selecting the bird list.
(`toriR`の出力ファイルは`csv`ファイルで、クリックとリスト選択された結果が出力されます。それらの情報は、"日にち、時刻、種名とクリックした周波数"で、1行毎に出力されます。))
```{bash}
Output file: 190501_230000-000000.csv
```

### Parameters of spectrogum extraction(スペクトログラム作成するための設定パラメータ)
```{bash}
# value is default.
fftsize=1024;#Fast Fourie Transfer size (512 or 1024)｜FFTサイズ
f_lcf=0000;           #Frequency of low cut filter[Hz]｜周波数の加減
f_hcf=12000;          #Frequency of high cut filter[Hz]｜周波数の上限
height_image="125";   #Image height of Spectrogram
ratio_clip="20";      #Clipping ratio of sound spectrogram generated
thrsd_clip_lo="-200"; #Threshold of clipping value of low side of spectrogram gererated
thrsd_clip_hi="-100"; #Threshold of clipping value of high side of spectrogram gererated
window_time="60";     #Spectrogram time window｜スペクトログラムの時間幅
width_image="624";    #Image width of Spectrogram
windows_a_page="4";   #Spectrograms a page｜1頁のあたりのスペクトログラム
```

## How to use `toriR` (`toriR`の使い方)
### Load source code 'toriR' (`toriR`コードを読む)
1. Open `RStudio`;      # `RStudio`を開く
2. Load 'toriR.R` code; # `toriR`コードを読み込む

### Set and edit parameters in the `toriR` source code; (`toriR`のソースコード無いのパラメータを編集する)
1. Set date parameter `date_analysis`: date_analysis <- 190501;(`date_analysis`を6桁の日にちを指定)
2. Set time parameter `time_analysis_sart` as a charactor: (`time_analysis`を文字列として開始時刻を指定する)
3. Edit candidates of bird's name of `spices` as a vector: (`spices`に鳥の声の種名を登録する。種名を""の中に入れる)
```{R}
spices <- c(
"play", "WAY｜save", "noise", "owl", "White's Thrush", "Japanese Green Pigeon"
);
```
4. Edit playing setting(option)(再生の設定(オプション))
```{R}
`volume` <- 8;# play volume;(再生音量の`volume`を設定)
`length_preplay` <- 1.; # length of playing time befor click(クリックより前の再生時間)
`length_play` <- 3;     # length of playing time after click(クリックの後の再生時間)
```

### Run `toriR`(トリルの実行)

1. Run of all `toriR` source code
   (`toriR`の全てのソースコードを実行する)
2. Click bird's voice on the spectrogram image
   (`toriR`のスペクトログラム画像上の鳥の声紋をクリック)
3. Select bird's name or play from list after moving focus on console)
   (コンソール上にフォーカスし、リストから鳥の名前を選ぶ)
4. When move to the next page, click the white area outside of the spectrogrum range on Plots or press `ESC` key
   (次のページに移動するにはPlots上のスペクトログラムの範囲外の白い部分をクリックするか`ESC`キーを押す)
5. When skip the rest of pages, press `ESC` immediately after clicking on the plot
   (残りのページをスキップしたい場合は、プロット上をクリックした直後に`Esc`キーを押します)


### Output file (出力ファイル)
190501_230000-000000.csv


## FAQ
1. [Q:Win]フォルダ名入力しCtrl+A→Ctrl+Enterを行ったところ文字化けしてエラーになってしまいました。
    - RStudioでtoriRを読み込ませて文字化けする場合は`Reopen with encording`で`UTF-8`を選んで呼んでください。


rstudio/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
[やってみたblog](https://www.softantenna.com/wp/software/keep-a-changeloag/)

## [0.9.0] - 2019-11-08

docker + Rstudio server for localhost 化

* [localhost only](http://localhost:8787) without auth
* [localhost/files/<REL_PATH>](http://localhost:8787/files/) でファイルアクセスはできるけど、directory browse はできない

## [0.83] - 2019-10-06

HO 公開用にコードを整える

1. 666形式のスペクトログラムを表示し、クリックしてリストから選択すると日にち,時刻,種名,周波数が出力される
2. 迷ったら即再生

## [0.8] - 2019-10-06

toriRをGitHub公開用に整える

TODO:  解析済みの画像ファイルは"190501_230000-000000_P01_toriR.png"などとなります

## [0.7] - 2019-09-23

toriRに改名。コードをシンプルにする。

## [0.5] - 2019-08-25

kikimimi_csvを~/Desktop/daybreak/original_birdlist/O大峰山とした

## [0.1] - 2019-06-20

新規作成


toriR0.81.R
# Select or linsten bird's voice on R
# @author woodiw2wopper <woodie2wopper@gmail.com>
# Copyright (C) 2019 woodiw2wopper
## ユーザ設定1｜日程と時刻の設定 ##############################
date_analysis      <- 190501
time_analysis_sart <- "040000";#文字列です。""で囲ってください

## ユーザ設定2｜ディレクトリの設定（ディレクトリは"/"で終わってください）#######
# Windows:
dir_HDD <- "C:/Users/hoge/";
# MacOS:
dir_HDD <- "/Users/osaka/toriR\ Dropbox/Sample\ Folder/toriR_demo/"
## ユーザ設定3｜鳥の名前 ########################################
spices <- c(
  "再生", "WAY｜保存", "雑音", "他の鳥",
  "フクロウ", "トラツグミ","ヨタカ", "ホトトギス",
  "ヤブサメS単上", "ヤブサメS単水", "ヤブサメS連続", "ヤブサメS駆足",
  "ミゾゴイ", "アオサギ", "ゴイサギ",
   "アカショウビン","ヤイロチョウ", "ウグイス","ハシブトガラス"
  );

## ユーザ設定4｜再生パラメータの設定 #############################
volume <- 8;          # play volume          ｜再生音量
length_preplay <- 1.; # play time befor click｜クリック前の再生時間
length_play <- 3;     # play time after click｜クリック後の再生時間

## ユーザ設定5｜ライブラリの読み込み #############################
#install.packages("png", dependencies = TRUE);
#install.packages("dplyr", dependencies = TRUE);
#install.packages("stringr", dependencies = TRUE);

# version########################################################
programname <- "toriR0.8.R"; # 191006 HO 公開用にコードを整える
# 機能1：666形式のスペクトログラムを表示し、クリックしてリストから選択すると
#        日にち,時刻,種名,周波数が出力される
# 機能2：迷ったら即再生
# [R0.8] 2019-10-06 toriRをGitHub公開用に整える
#（これまだできていない）        解析済みの画像ファイルは"190501_230000-000000_P01_toriR.png"などとなります
# [R0.7] 2019-09-23 toriRに改名。コードをシンプルにする。
# [R0.5] 2019-08-25 kikimimi_csvを~/Desktop/daybreak/original_birdlist/O大峰山とした
# [R0.1] 2019-06-20 新規作成

# R script from here ###########################################
os <- strsplit(osVersion, " ")[[1]][1];
dir_dist <- paste0(dir_HDD, date_analysis);

if(as.numeric(time_analysis_sart) < 180000){
  date_analysis <- format((as.Date(as.character(date_analysis), "%y%m%d") + 1) , "%y%m%d")
# }else{
#   dir_HDD <- paste0(dir_HDD, format((as.Date(as.character(date_analysis), "%y%m%d") ) , "%y%m%d"))
}
# dir_HDD <- paste0(dir_HDD, format((as.Date(as.character(date_analysis), "%y%m%d") + 1) , "%y%m%d"))


time_analysis <- paste(date_analysis, as.character(time_analysis_sart), sep="_");
time_start    <- strptime(time_analysis, format="%y%m%d_%H%M%S");
file_param    <- paste0(format(time_start, "%y%m%d_%H%M%S-"), format(time_start + 3600, "%H%M%S"), ".psgrm");

setwd(dir_dist)
load(file = file_param)

library(png)
library(dplyr)
library(stringr)

str_play_trim <- function(os, volume, file_song, second_start=0, length_play=5){
  if (os == "Windows" ) {
    res <- sprintf("sox -V0 -v %d %s -t waveaudio trim %s %d", volume, file_song, second_start, length_play)
  }
  if (os == "macOS"){
    res <- sprintf("play -V0 -v %d %s trim %s %d", volume, file_song, second_start, length_play)
  }
  return(res)
}

str_play <- function(os, volume, file_song){
  if (os == "Windows" ) {
    res <- sprintf("sox -V0 -v %d %s -t waveaudio", volume, file_song)
  }
  if (os == "macOS"){
    res <- sprintf("play -V0 -v %d %s", volume, file_song)
  }
  return (res)
}


# file setting ######
#inputfile <- file_param
body_file_param <- stringr::str_remove(file_param, "\\..+$")
file_wav        <- stringr::str_replace(file_param, ".psgrm", ".wav")
csvfile_output  <- stringr::str_replace(file_param, ".psgrm", ".csv")
csvfile_output  <- paste0(dir_dist, "/", csvfile_output)


#if( file.exists(csvfile_output) ) {
#  txt_alarm <- str_play(os, 1, file_song = "../file_exist.wav")
#  system(txt_alarm)
#  cat("Caution: The csv file exist.")
#}

pngs <- list.files(pattern = paste0(body_file_param, "_P[0-9][0-9].png"))
num_png <- length(pngs)

# Recordint date and time ####
file_info <- unlist(stringr::str_split(body_file_param, "[_-]")); # unlistが必要。最初の3つを取り出す
record_date <- file_info[1]; # 録音日
time_start  <- paste(record_date, file_info[2], sep =" ")
time_start  <- strptime(time_start, format="%y%m%d %H%M%S") ; # 
time_end    <- paste(record_date, file_info[3], sep = " "); # 録音終了時間
time_end    <- strptime( time_end, format="%y%m%d %H%M%S") ; # 録音開始時刻

now <- Sys.time()
txt_kikimimi <- c(
  paste0("# プログラム:programname=", programname, "\n", 
         "# 解析開始時間:StartTime_toriR=", now, "\n",
         format(time_start,"%Y.%m.%d,%H:%M:%S,R,,\n")
  )
)
cat(txt_kikimimi, file = csvfile_output, append = TRUE)

# windowの表示画面上の場所指定 ####
if (windows_a_page == 4){
  xl <- 0.044; xr <- 0.93;  
  yb <- 0.805; yt <- 0.978;
  dy <- 0.25;
}


#表示の設定####
if (os == "macOS") {
  par(family = "HiraKakuProN-W3")
}
if (os == "windows"){
  windowsFonts(MEI = windowsFont("Meiryo"))
  par(family = MEI)
}
images <- list(1:num_png)
for (num_page in 1:num_png){
  images[[num_page]] <- readPNG(pngs[num_page])
}
#toriRの結果データベース
db <- data.frame(
  num_png=integer(),
  x=numeric(),
  y=numeric(),
  label=character(),
  col_text=character()
)
db <- tribble(
  ~num_png, ~x, ~y, ~label,~col_text,
  1,        0,   0,  "temp", "white"
)
for (num_page in 1:num_png){
# for (num_page in 1:2){
  flag_page_end <- FALSE;
  pngfile_input  <- pngs[num_page];
  cat("R> num_page = ",num_page,"\n")
  cat("R> ", pngfile_input, "\n")
  cat(paste0("# スペクトログラム:spectrogram=", pngfile_input, "\n"), file = csvfile_output, append =TRUE)
  num_page <-   stringr::str_sub(pngfile_input, start = 23, end = 24) %>% as.numeric()
  par(mar=c(0, 0, 0, 0)); #外側のマージンをゼロにする
  plot(NULL, xlim = c(0,1), ylim = c(0, 1), xlab = "", ylab = "", yaxt = "n", xaxt = "n",xaxs = 'i', yaxs = 'i')
  rasterImage(images[[num_page]], 0, 0, 1, 1)
  for( i in 0:(windows_a_page-1)){
    rect( xl, yb - i * dy, xr, yt - i * dy, col="transparent", border = "red")
  }
  while (flag_page_end == FALSE){
    cat("R> 鳥声をクリック。なければ範囲外をクリック。またはESPキー:");
    z <- locator(n=1); # %>% as.data.frame;
    if (is.null(z) == TRUE){# ESCが押されたなら
      flag_page_end <- TRUE
      break;
    }
    x <- z$x; 
    y <- z$y; 
    y_position <- "out_of_range"
    for( k in 0:(windows_a_page - 1)) {# locator()の場所判定
      if(yb - k * dy <= y  && y < yt - k * dy) {y_position <- k}
    }
    if(y_position == "out_of_range"){break};# 範囲外選択
    points(x, y, col="RED", pch = 20);
    col_text <- "RED"; #抽出種名の文字の色
    
    time_offset <- (num_page - 1) * window_time * windows_a_page + window_time * y_position
    second_locator <- (time_offset + (x - xl) * (window_time)/(xr - xl)) 
    if(second_locator < 0)(length_preplay <- 0)
    txt_play <- str_play_trim(os, volume, file_wav, second_locator - length_preplay, length_play)
    time_locator   <- (time_offset + (x - xl) * (window_time)/(xr - xl) + time_start) %>% format("%Y.%m.%d,%H:%M:%S")
    freq_locator   <- (f_lcf + (y - (yb - y_position * dy)) * (f_hcf - f_lcf)/(yt - yb))
    
    answer <- menu(spices, title="\nR> 種類を選択してください:")
    # cat(paste0("Select:", spices[answer]))
    while(spices[answer] == "再生") {#再生
      cat (txt_play);
      cat (paste0("# 再生:", txt_play, "\n"), file = csvfile_output, append = TRUE);
      
      system(txt_play);
      answer <- menu(spices, title="R> 種類を再選択してください");
      cat(paste0("Reselection = ", spices[answer],"\n"))
    }
    if(spices[answer] ==  "WAY｜保存") {#
      time_locator_file  <- (time_offset + (x - xl) * (window_time)/(xr - xl) + time_start) %>% format("%y%m%d_%H%M%S")
      file_sox_save <- paste0(time_locator_file, "｜WAY.wav")
      txt_sox_save  <- sprintf("sox -v %d %s %s trim %s %d", volume, file_wav, file_sox_save, second_locator - length_preplay, length_play)
      system(txt_sox_save);
      col_text <- "YELLOW"
    }
    if (answer == "0") {
      flag_page_end <- TRUE;
      break;
    }else{
      txt_kikimimi <- sprintf("%s,%s,F=%d[Hz]\n", time_locator , spices[answer], as.integer(freq_locator))
      cat(paste0("\nkikimimi>",txt_kikimimi))
      cat(txt_kikimimi, file = csvfile_output, append = TRUE)
      db <- rbind(db,c(num_page, x, y, spices[answer], col_text))
      text(x, y + 0.03, label = spices[answer], col= col_text); # 選択した文字列を図中に表示

    }
  }
  if(num_page == num_png){
    dev.off()
    before <- now
    now <- Sys.time()
    txt_ending <- paste0(
      format(time_end,"%Y.%m.%d,%H:%M:%S,STOP,,\n"),
      "# 終了解析時間:StopTime_toriR=", now, "\n",
      "# 分析時間:AnalysisTime_toriR=", format(as.numeric(difftime(now, before, unit="min")), digit=2), "min.\n"
      )
    cat(txt_ending, file = csvfile_output, append = TRUE)
    cat(txt_ending)
  }
}

##画像に抽出結果を出力する ####

db <- db[-1,];
for (num_page in 1:num_png){
# for (num_page in 1:2){
  pngfile_input  <- pngs[num_page];
  pngfile_output <- stringr::str_replace(pngfile_input, ".png", "_toriR.png");
  pngfile_output <- pngfile_input;
  png(pngfile_output, width = width_image, height = windows_a_page * height_image, bg="white")
  par(mar=c(0, 0, 0, 0)); #外側のマージンをゼロにする
  plot(NULL, xlim = c(0,1), ylim = c(0, 1), xlab = "a", ylab = "", yaxt = "n", xaxt = "n",xaxs = 'i', yaxs = 'i')
  rasterImage(images[[num_page]], 0, 0, 1, 1)
  text(0.9,0.99,label="toriR",col="RED")
  for (ii in 1:nrow(db)){
    w <- db[ii,];
    if ( num_page == w$num_png ){
      points(w$x, w$y, col=w$col_text, pch = 20);
      text(as.numeric(w$x), as.numeric(w$y) + 0.03, label = w$label, col= w$col_text);
    }
  }
  dev.off()
}
